src/coreclr/jit/lclmorph.cpp

```diff
@@ -936,18 +936,45 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
                 var_types elementType = indir->TypeGet();
                 lclNode               = indir->gtGetOp1()->BashToLclVar(m_compiler, lclNum);
 
-                if (elementType == TYP_FLOAT)
+                switch (elementType)
                 {
-                    GenTree* indexNode = m_compiler->gtNewIconNode(offset / genTypeSize(elementType));
-                    hwiNode = m_compiler->gtNewSimdGetElementNode(elementType, lclNode, indexNode, CORINFO_TYPE_FLOAT,
-                                                                  genTypeSize(varDsc));
-                }
-                else
-                {
-                    assert(elementType == TYP_SIMD12);
-                    assert(genTypeSize(varDsc) == 16);
-                    hwiNode = m_compiler->gtNewSimdHWIntrinsicNode(elementType, lclNode, NI_Vector128_AsVector3,
-                                                                   CORINFO_TYPE_FLOAT, 16);
+                    case TYP_FLOAT:
+                    {
+                        GenTree* indexNode = m_compiler->gtNewIconNode(offset / genTypeSize(elementType));
+                        hwiNode            = m_compiler->gtNewSimdGetElementNode(elementType, lclNode, indexNode,
+                                                                      CORINFO_TYPE_FLOAT, genTypeSize(varDsc));
+                        break;
+                    }
+                    case TYP_SIMD12:
+                    {
+                        assert(genTypeSize(varDsc) == 16);
+                        hwiNode = m_compiler->gtNewSimdHWIntrinsicNode(elementType, lclNode, NI_Vector128_AsVector3,
+                                                                       CORINFO_TYPE_FLOAT, 16);
+                        break;
+                    }
+                    case TYP_SIMD8:
+#if defined(FEATURE_SIMD) && defined(TARGET_XARCH)
+                    case TYP_SIMD16:
+                    case TYP_SIMD32:
+#endif
+                    {
+                        assert(genTypeSize(elementType) * 2 == genTypeSize(varDsc));
+                        if (offset == 0)
+                        {
+                            hwiNode = m_compiler->gtNewSimdGetLowerNode(elementType, lclNode, CORINFO_TYPE_FLOAT,
+                                                                        genTypeSize(varDsc));
+                        }
+                        else
+                        {
+                            assert(offset == genTypeSize(elementType));
+                            hwiNode = m_compiler->gtNewSimdGetUpperNode(elementType, lclNode, CORINFO_TYPE_FLOAT,
+                                                                        genTypeSize(varDsc));
+                        }
+
+                        break;
+                    }
+                    default:
+                        unreached();
```

Comment:
Are we confident that `unreached()` is "correct" here or can/should we simply bail out of the transformation?

The main consideration would probably be that `Vector<T>` and `Vector64<T>` are backed by `ulong` fields. That shouldn't normally be an issue, but with all the various `Unsafe.As/BitCast` and other tricks, I'm not positive its not possible to see something here still.

Reply:
I think it's fine -- `SelectLocalIndirTransform` only returns `WithElement`/`GetElement` for the cases that we can handle here.